### DISCUSSION TITLE
Accessing player

### DISCUSSION THREAD

**Discussion by AndrejkoL** [Annoraaq/grid-engine#101] [Q&A]
Hi Annoraaq. I started using your plugin and grid system and everything is great but i dont know how to acces player as object. Like if i wanted to add collider for example:
 this.physics.add.collider(player, worldLayer);
But player is just undefined and i can't acces is through id either. I just need to know how to access player from this
const gridEngineConfig = {
    characters: [
      {
        id: "player",
        sprite: playerSprite,..
Thanks

**Comment by Annoraaq** [comment]
Hi Andrej,

great to hear you like the plugin :)

So in your concrete example, you should simply use `playerSprite`, that you pass to the plugin. This should be a `Phaser.GameObjects.Sprite` which is also a `Phaser.GameObjects.GameObject` and therefore can be passed to [collider(...)](https://photonstorm.github.io/phaser3-docs/Phaser.Physics.Arcade.Factory.html#collider__anchor)

Does that help or do you mean something else by "player"?
The plugin internally just uses an internal class wrapping the sprite and that should not be used outside of the plugin.

Or is there any information about the player that you would like to access from outside? If so, what concrete information is that?

**Reply by AndrejkoL** [reply]
Thanks for replying. PlayerSprite wasn't doing anything either so I thought I need to access it through id somehow. Turns out when I changed this code from the guide:
 const playerSprite = this.add.sprite(0, 0, "player");
to:
 const playerSprite = this.physics.add.sprite(0, 0, "player");
It actually all works fine. Thanks for making me check sprites. Not sure if it is intended in guide or not but i spend whole day on it :)

**Reply by Annoraaq** [reply]
Now you are probably using [Phaser.Physics.Arcade.Sprite](https://photonstorm.github.io/phaser3-docs/Phaser.Physics.Arcade.Sprite.html). This is fine for the plugin but you will have to be careful because now you are using two independent physic engines on the same sprite. One for grid based collision and one for arcade physics collisions.

This can lead to unexpected behavior of the plugin if it can't move the sprite because of a collision caused by the arcade physics. An example for a problem is the collideWordBounds property:

https://github.com/Annoraaq/grid-engine/issues/26

Please let me know if you face any problems and maybe some can be prevented by the plugin.

**Reply by AndrejkoL** [reply]
How would you create object then which you can collide or interact with? If i create something like this:
 this.boxGroup = this.physics.add.staticGroup();
 this.boxGroup.get(64,64, "object", 10)
 .setSize(32,32);
Then my code for collision does not work:
 this.physics.add.collider(playerSprite, this.boxGroup);
I'm just bit stuck on this part but without creating objects to interact with i can't really do anything. Any help would be appreciated

**Reply by Annoraaq** [reply]
What is the expected behavior that you would like to see?

Is it that you want the player to stop moving if it collides with a certain object?

This is easy in arcade physics but it causes problems in "grid physics". In a grid world the player always has a tile position. If he stops between two tiles because of such a collision, on which tile position will he be?

Therefore in a grid world the collisions are usually tile-based. That means that you are not checking for collisions like in an arcade way. Moreover collision detection is done via checking if the tile the player wants to move to is currently blocked by a colliding object or not. So everything is about tile positions and not pixel positions. That is the big difference to the arcade physics.

You can still use the pixel position of the player to trigger interactions. For example if you want to shoot bullets to players, you can check for an overlap between the bullet and the player sprite.

But for the above reason the movement can't be stopped by such an object because that would conflict with the tile based movement. If you need such a functionality it might be better to completely use arcade physics and switch to a pixel based movement.

**Comment by AndrejkoL** [comment]
I was actually looking for a way to interact with objects on the tile next
to me but still having difficulties with it because before i was using
collider to call a function for something to happen. It's just a little
hiccup im sure i will eventually figure it out :)

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Reply by AndrejkoL** [reply]
Also just last question please. How would you create objects with grid colliding physics and set some custom start position like you have for player:
  const gridEngineConfig = {
    characters: [
      {
        id: "player",
        sprite: playerSprite,
        walkingAnimationMapping: 6,
        startPosition: new Phaser.Math.Vector2(8, 8),
      },
    ],
  };
I tried adding objects[] to gridEngineConfig but start position does not work in it. Thanks

**Reply by Annoraaq** [reply]
Ok I think I see your problem now. Let's work out a solution.

You could just add the objects to the plugin and pretend they are characters. This would enable grid physics collisions and you could get their tile position with the 'getPosition()` method of the plugin. You might want to turn off the walking animation for those objects (Oops, it's missing in the docs 😨 . I will add it)

Here is an example:

```Javascript
const gridEngineConfig = {
  characters: [
    {
      id: "player",
      sprite: playerSprite,
      walkingAnimationMapping: 6,
      startPosition: new Phaser.Math.Vector2(8, 8),
    },
    {
      id: "myObject",
      sprite: objectSprite,
      walkingAnimationEnabled: false,
      startPosition: new Phaser.Math.Vector2(4, 4),
    }
  ],
};
```

However, this would require that the "objects" must also be moved in a grid physics fashion (only tile based movement) and further the objects would need to be a Phaser.GameObjects.Sprite or a subclass of that.

Additionally it would probably help if we add a `setTilePosition(...)` method to the plugin. So you can freely reposition your object after adding it.

Does that help somehow?

Edit:
You can then use the `getFacingDirection()` and `getPosition(...)` methods of the plugin to check whether your char is looking towards the object and whether the object is on a neighboring tile.

Also one more hint: If your object is more complex than a sprite, that is fine. Just pass the sprite to the plugin and keep a mapping somewhere between the "characterId" you passed to the plugin and your object reference.

**Comment by AndrejkoL** [comment]
Ye actually that would be awesome. But it works as you suggested. Now with
getPosition i should be able to do function somehow  to interact with those
objects so i think all set. Thank you

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by borkowsky** [comment]
I get character by following code:
```
let character = this.gridEngine.gridCharacters.get('char-id')

```

**Reply by Annoraaq** [reply]
While this is possible I would not recommend it. In the typescript code the `gridCharacters` map is a private property. So it is not part of the public API and can therefore change with any update.

Let me know what functionality of the character you need. Then I can create a public API method for this that is stable. The character you are getting out of the private `gridCharacters` map is only intended for internal use in the plugin itself.

**Reply by borkowsky** [reply]
I think ```getCharacter(id)``` method should be useful for do that

**Reply by Annoraaq** [reply]
Yeah but that method should not return the internally used `GridCharacter` object. It would be interesting to know why you need the internal character object and then I could expose this functionality in a cleaner way.

**Reply by borkowsky** [reply]
> Yeah but that method should not return the internally used `GridCharacter` object. It would be interesting to know why you need the internal character object and then I could expose this functionality in a cleaner way.

I need it to access player sprite or player container to add / modify / remove container's children

**Reply by Annoraaq** [reply]
I see. But you are creating the sprite/container outside of the plugin. So a cleaner way would be to keep track of your sprites outside of it as well.

However, I can additionally add a clean way for retrieving the sprite that is currently used by the plugin for a certain character.

**Reply by borkowsky** [reply]
> I see. But you are creating the sprite/container outside of the plugin. So a cleaner way would be to keep track of your sprites outside of it as well.
> 
> However, I can additionally add a clean way for retrieving the sprite that is currently used by the plugin for a certain player.

You're right. I should keep references to containers at another place. I have map of players, stored in scene variable, so i need to add reference to containers on map.
